File: src/SimpleLevelEditor.Formats.EntityConfig/EntityPropertyTypeDescriptor.cs
using Dunet;
using SimpleLevelEditor.Formats.Core;
using System.Diagnostics;
using System.Numerics;

namespace SimpleLevelEditor.Formats.EntityConfig;

[Union]
public partial record EntityPropertyTypeDescriptor
{
	public sealed partial record BoolProperty(bool Default);
	public sealed partial record IntProperty(int Default, int? Step, int? Min, int? Max);
	public sealed partial record FloatProperty(float Default, float? Step, float? Min, float? Max);
	public sealed partial record Vector2Property(Vector2 Default, float? Step, float? Min, float? Max);
	public sealed partial record Vector3Property(Vector3 Default, float? Step, float? Min, float? Max);
	public sealed partial record Vector4Property(Vector4 Default, float? Step, float? Min, float? Max);
	public sealed partial record StringProperty(string Default);
	public sealed partial record RgbaProperty(Rgba Default);
	public sealed partial record RgbProperty(Rgb Default);

	// public EntityPropertyValue GetDefaultValue()

	public float GetStep()
	{
		return this switch
		{
			IntProperty p => p.Step ?? 0,
			FloatProperty p => p.Step ?? 0,
			Vector2Property p => p.Step ?? 0,
			Vector3Property p => p.Step ?? 0,
			Vector4Property p => p.Step ?? 0,
			_ => 0,
		};
	}

	public float GetMin()
	{
		return this switch
		{
			IntProperty p => p.Min ?? 0,
			FloatProperty p => p.Min ?? 0,
			Vector2Property p => p.Min ?? 0,
			Vector3Property p => p.Min ?? 0,
			Vector4Property p => p.Min ?? 0,
			_ => 0,
		};
	}

	public float GetMax()
	{
		return this switch
		{
			IntProperty p => p.Max ?? 0,
			FloatProperty p => p.Max ?? 0,
			Vector2Property p => p.Max ?? 0,
			Vector3Property p => p.Max ?? 0,
			Vector4Property p => p.Max ?? 0,
			_ => 0,
		};
	}

	public string GetTypeId()
	{
		return this switch
		{
			BoolProperty => "Bool",
			IntProperty => "Int",
			FloatProperty => "Float",
			Vector2Property => "Vector2",
			Vector3Property => "Vector3",
			Vector4Property => "Vector4",
			StringProperty => "String",
			RgbaProperty => "Rgba",
			RgbProperty => "Rgb",
			_ => throw new UnreachableException(),
		};
	}

	public Vector4 GetDisplayColor()
	{
		return this switch
		{
			BoolProperty => new Vector4(0.00f, 0.25f, 1.00f, 1.00f),
			IntProperty => new Vector4(0.00f, 0.50f, 1.00f, 1.00f),
			FloatProperty => new Vector4(0.00f, 0.70f, 0.00f, 1.00f),
			Vector2Property => new Vector4(0.00f, 0.80f, 0.00f, 1.00f),
			Vector3Property => new Vector4(0.00f, 0.90f, 0.00f, 1.00f),
			Vector4Property => new Vector4(0.00f, 1.00f, 0.00f, 1.00f),
			StringProperty => new Vector4(1.00f, 0.50f, 0.00f, 1.00f),
			RgbaProperty => new Vector4(1.00f, 0.75f, 0.00f, 1.00f),
			RgbProperty => new Vector4(1.00f, 1.00f, 0.00f, 1.00f),
			_ => throw new UnreachableException(),
		};
	}

	public EntityPropertyTypeDescriptor DeepCopy()
	{
		return this switch
		{
			BoolProperty p => new BoolProperty(p.Default),
			IntProperty p => new IntProperty(p.Default, p.Step, p.Min, p.Max),
			FloatProperty p => new FloatProperty(p.Default, p.Step, p.Min, p.Max),
			Vector2Property p => new Vector2Property(p.Default, p.Step, p.Min, p.Max),
			Vector3Property p => new Vector3Property(p.Default, p.Step, p.Min, p.Max),
			Vector4Property p => new Vector4Property(p.Default, p.Step, p.Min, p.Max),
			StringProperty p => new StringProperty(p.Default),
			RgbaProperty p => new RgbaProperty(p.Default),
			RgbProperty p => new RgbProperty(p.Default),
			_ => throw new UnreachableException(),
		};
	}
}

File: src/SimpleLevelEditor.Formats/Types/EntityConfig/EntityPropertyTypeDescriptor.fs
namespace SimpleLevelEditor.Formats.Types.EntityConfig

open System.Numerics
open SimpleLevelEditor.Formats.Types
open SimpleLevelEditor.Formats.Types.Level

type EntityPropertyTypeDescriptor =
    | BoolProperty    of Default: bool
    | IntProperty     of Default: int     * Step: Option<int>     * Min: Option<int>     * Max: Option<int>
    | FloatProperty   of Default: float32 * Step: Option<float32> * Min: Option<float32> * Max: Option<float32>
    | Vector2Property of Default: Vector2 * Step: Option<float32> * Min: Option<float32> * Max: Option<float32>
    | Vector3Property of Default: Vector3 * Step: Option<float32> * Min: Option<float32> * Max: Option<float32>
    | Vector4Property of Default: Vector4 * Step: Option<float32> * Min: Option<float32> * Max: Option<float32>
    | StringProperty  of Default: string
    | RgbProperty     of Default: Rgb
    | RgbaProperty    of Default: Rgba

    member this.DefaultValue =
        match this with
        | BoolProperty    value            -> EntityPropertyValue.Bool    value
        | IntProperty     (value, _, _, _) -> EntityPropertyValue.Int     value
        | FloatProperty   (value, _, _, _) -> EntityPropertyValue.Float   value
        | Vector2Property (value, _, _, _) -> EntityPropertyValue.Vector2 value
        | Vector3Property (value, _, _, _) -> EntityPropertyValue.Vector3 value
        | Vector4Property (value, _, _, _) -> EntityPropertyValue.Vector4 value
        | StringProperty  value            -> EntityPropertyValue.String  value
        | RgbProperty     value            -> EntityPropertyValue.Rgb     value
        | RgbaProperty    value            -> EntityPropertyValue.Rgba    value

    // TODO: Return Option<float32> instead of float32
    member this.Step : float32 =
        match this with
        | IntProperty     (_, step, _, _) -> if step.IsSome then float32 step.Value else 0f
        | FloatProperty   (_, step, _, _) -> if step.IsSome then step.Value else 0f
        | Vector2Property (_, step, _, _) -> if step.IsSome then step.Value else 0f
        | Vector3Property (_, step, _, _) -> if step.IsSome then step.Value else 0f
        | Vector4Property (_, step, _, _) -> if step.IsSome then step.Value else 0f
        | _ -> 0f

    // TODO: Return Option<float32> instead of float32
    member this.MinValue : float32 =
        match this with
        | IntProperty     (_, _, min, _) -> if min.IsSome then float32 min.Value else 0f
        | FloatProperty   (_, _, min, _) -> if min.IsSome then min.Value else 0f
        | Vector2Property (_, _, min, _) -> if min.IsSome then min.Value else 0f
        | Vector3Property (_, _, min, _) -> if min.IsSome then min.Value else 0f
        | Vector4Property (_, _, min, _) -> if min.IsSome then min.Value else 0f
        | _ -> 0f

    // TODO: Return Option<float32> instead of float32
    member this.MaxValue : float32 =
        match this with
        | IntProperty     (_, _, _, max) -> if max.IsSome then float32 max.Value else 0f
        | FloatProperty   (_, _, _, max) -> if max.IsSome then max.Value else 0f
        | Vector2Property (_, _, _, max) -> if max.IsSome then max.Value else 0f
        | Vector3Property (_, _, _, max) -> if max.IsSome then max.Value else 0f
        | Vector4Property (_, _, _, max) -> if max.IsSome then max.Value else 0f
        | _ -> 0f

    member this.GetTypeId() =
        match this with
        | BoolProperty    _ -> nameof Bool
        | IntProperty     _ -> nameof Int
        | FloatProperty   _ -> nameof Float
        | Vector2Property _ -> nameof Vector2
        | Vector3Property _ -> nameof Vector3
        | Vector4Property _ -> nameof Vector4
        | StringProperty  _ -> nameof String
        | RgbProperty     _ -> nameof Rgb
        | RgbaProperty    _ -> nameof Rgba

    member this.GetDisplayColor() =
        match this with
        | BoolProperty    _ -> new Vector4(0.00f, 0.25f, 1.00f, 1.00f)
        | IntProperty     _ -> new Vector4(0.00f, 0.50f, 1.00f, 1.00f)
        | FloatProperty   _ -> new Vector4(0.00f, 0.70f, 0.00f, 1.00f)
        | Vector2Property _ -> new Vector4(0.00f, 0.80f, 0.00f, 1.00f)
        | Vector3Property _ -> new Vector4(0.00f, 0.90f, 0.00f, 1.00f)
        | Vector4Property _ -> new Vector4(0.00f, 1.00f, 0.00f, 1.00f)
        | StringProperty  _ -> new Vector4(1.00f, 0.50f, 0.00f, 1.00f)
        | RgbProperty     _ -> new Vector4(1.00f, 0.75f, 0.00f, 1.00f)
        | RgbaProperty    _ -> new Vector4(1.00f, 1.00f, 0.00f, 1.00f)

    member this.DeepCopy() =
        match this with
        | BoolProperty    value                   -> BoolProperty    value
        | IntProperty     (value, step, min, max) -> IntProperty     (value, step, min, max)
        | FloatProperty   (value, step, min, max) -> FloatProperty   (value, step, min, max)
        | Vector2Property (value, step, min, max) -> Vector2Property (value, step, min, max)
        | Vector3Property (value, step, min, max) -> Vector3Property (value, step, min, max)
        | Vector4Property (value, step, min, max) -> Vector4Property (value, step, min, max)
        | StringProperty  value                   -> StringProperty  value
        | RgbProperty     value                   -> RgbProperty     value
        | RgbaProperty    value                   -> RgbaProperty    value
